File: list/list.py
class List:

    def __getitem__(self, n):
        return reduce(lambda (result, n), value: (result or value, n - 1) if n <= 0 else (result, n - 1), self, (None, n))[0]

    def __len__(self):
        return reduce(lambda n, _: n + 1, self, 0)

    def reverse(self):
        return reduce(lambda tail, head: Cons(head, tail), self, Nil())

    @staticmethod
    def create(*values):
        return reduce(lambda acc, head: lambda k: acc(lambda tail: Cons(head, k(tail))), values, lambda k: k(Nil()))(lambda x: x)

class Nil(List):

    def __add__(self, other):
        return other

    def __eq__(self, other):
        return isinstance(other, Nil)

    def __iter__(self):
        raise StopIteration()

class Cons(List):

    def __init__(self, head, tail):
        self.head = head
        self.tail = tail

    def __add__(self, other):
        return Cons(self.head, self.tail + other)

    def __eq__(self, other):
        return isinstance(other, Cons) and self.head == other.head and self.tail == other.tail

    def __iter__(self):
        yield self.head
        for value in self.tail:
            yield value

if __name__ == '__main__':
    foo = List.create(0, 1, 2)
    assert len(foo) == 3
    assert foo[1] == 1
    assert foo[2] == 2
    assert foo[3] == None
    assert foo + foo == List.create(0, 1, 2, 0, 1, 2)
    assert foo.reverse() == List.create(2, 1, 0)

File: list/list.rb
class List

  include Enumerable

  def reverse
    inject(Nil.new) { |tail, head| Cons.new(head, tail) }
  end

  def [](n)
    inject([nil, n]) { |acc, value| if acc[1] <= 0 then [acc[0] || value, acc[1] - 1] else [acc[0], acc[1] - 1] end }[0]
  end

  def self.create(*values)
    values.inject(->(k) { k.call(Nil.new) }) { |acc, head|
      ->(k) {
        acc.call(->(tail) { Cons.new(head, k.call(tail)) })
      }
    }.call(->(x) { x })
  end

end

class Nil < List

  def +(list)
    list
  end

  def ==(list)
    list.is_a?(Nil)
  end

  def each(&block)
  end

end

class Cons < List

  attr_reader :head, :tail

  def initialize(head, tail)
    @head = head
    @tail = tail
  end

  def +(list)
    Cons.new(@head, @tail + list)
  end

  def ==(list)
    list.is_a?(Cons) && @head == list.head && @tail == list.tail
  end

  def each(&block)
    block.call(@head)
    @tail.each { |x| block.call(x) }
  end

end

if __FILE__ == $0
  require 'test/unit/assertions'
  include Test::Unit::Assertions
  hoge = List.create(0, 1, 2)
  assert(hoge.count == 3)
  assert(hoge[1] == 1)
  assert(hoge[2] == 2)
  assert(hoge[3] == nil)
  assert(hoge + hoge == List.create(0, 1, 2, 0, 1, 2))
  assert(hoge.reverse() == List.create(2, 1, 0))
end
